### PR TITLE
Hive 25518 compaction txn handle NPE fix

### PR DESCRIPTION
**Pull request by vcsomor**
In the hive-exec when the `Worker` in the Compaction encounters an Exception it might try to mark failed a compaction without any compaction info. To make the method more resilient, a null-safe code branch was added to the int the `CompactionTxnHandler`

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 1]
```diff
@@ -1104,89 +1104,101 @@ public boolean checkFailedCompactions(CompactionInfo ci) throws MetaException {
...
         LOG.debug("Going to commit");
         closeStmt(pStmt);
         dbConn.commit();
       } catch (SQLException e) {
-        LOG.warn("markFailed(" + ci.id + "):" + e.getMessage());
+        LOG.warn("markFailed(" + ((ci != null) ? ci.id : "missing id") + "):" + e.getMessage());
         LOG.debug("Going to rollback");
         rollbackDBConn(dbConn);
         checkRetryable(dbConn, e, "markFailed(" + ci + ")");
         LOG.error("markFailed(" + ci + ") failed: " + e.getMessage(), e);
```
Comment (klcopp) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1193]: These will call ci#toString, right?

Comment (vcsomor) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1193]: Not really.
If the `ci == null` then it just adds nullm otherwise the `toString()` getting triggered.
```java
    CompactionInfo ci = null;
    System.out.println("(" + ci + ")");

    ci = new CompactionInfo("x", "y", "z", CompactionType.MINOR);
    System.out.println("(" + ci + ")");
```
```
(null)
(id:0,dbname:x,...)
```

File: ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestWorker.java [hunk 2]
```diff
@@ -39,19 +39,15 @@
...
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.io.ByteArrayInputStream;
-import java.io.ByteArrayOutputStream;
-import java.io.DataInput;
-import java.io.DataInputStream;
-import java.io.DataOutput;
-import java.io.DataOutputStream;
+import java.io.*;
```
[the remark this answers is not in the document]
Comment (vcsomor) [reply ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestWorker.java:50]: I'll change it back and I'll adjust the IDEA settings to avoid this in the future

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java (hunk 1 above)
Comment (klcopp) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1193]: Got it, thanks!

**Review by klcopp — APPROVED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 3]
```diff
@@ -585,6 +584,12 @@ protected Boolean findNextCompactionAndExecute(boolean computeStats) {
     return true;
   }
 
+  @VisibleForTesting
+  public CompactionInfo nextCompactForWorkerAndRuntime() throws TException {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:588]: could we please rename it to 'getNextCompactRequest'

Comment (vcsomor) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:588]: It is not really a getter, though. But I can do this minor change.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 4]
```diff
@@ -3487,6 +3487,16 @@ public HeartbeatTxnRangeResponse heartbeatTxnRange(HeartbeatTxnRangeRequest rqst
     }
   }
 
+  long generateCompactionQueueId(Connection dbConn) throws SQLException, MetaException {
+    Statement stmt = null;
+    try {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:3492]: use try-with-resources statement so you don't need to manage close

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 5]
```diff
@@ -1104,89 +1104,101 @@ public boolean checkFailedCompactions(CompactionInfo ci) throws MetaException {
...
-            throw new IllegalStateException("No record with CQ_ID=" + ci.id + " found in COMPACTION_QUEUE");
-          }
-        }
-        if(ci.id == 0) {
-          //The failure occurred before we even made an entry in COMPACTION_QUEUE
-          //generate ID so that we can make an entry in COMPLETED_COMPACTIONS
-          ci.id = generateCompactionQueueId(stmt);
-          //mostly this indicates that the Initiator is paying attention to some table even though
-          //compactions are not happening.
+        if(ci == null) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1116]: nit: space

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 6]
```diff
@@ -1104,89 +1104,101 @@ public boolean checkFailedCompactions(CompactionInfo ci) throws MetaException {
...
+          LOG.debug("CompactionInfo was null, creating a new one");
+          ci  = new CompactionInfo();
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1118]: Why is this needed? Are we going to process this compaction queue entry?

Comment (vcsomor) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1118]: It denotes something wrong is going on... As discussed with @klcopp probably we keep this PR on shelve

Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:1118]: my two cents here: I think, it would make sense to return right away from markFailed if a `NULL` ci is passed.
